reports/tasks/msisdn_message_report.py

```diff
@@ -13,63 +16,200 @@ class GenerateMSISDNMessageReport(BaseTask):
     MSISDNs.
     """
 
-    def run(self, task_status_id, msisdns=[]):
+    def run(self, start_date, end_date, task_status_id, msisdns=[],
+            email_recipients=[], email_sender=settings.DEFAULT_FROM_EMAIL,
+            email_subject='Seed Control Interface Generated Report', **kwargs):
+
+        start_date = datetime.strptime(start_date, "%Y-%m-%d")
+        end_date = datetime.strptime(end_date, "%Y-%m-%d")
+
         task_status = ReportTaskStatus.objects.get(id=task_status_id)
         task_status.status = ReportTaskStatus.RUNNING
         task_status.save()
 
-        spreadsheet = self.create_spreadsheet(msisdns)
+        is_client = IdentityStoreApiClient(
+            settings.IDENTITY_STORE_TOKEN,
+            settings.IDENTITY_STORE_URL,
+        )
+        ms_client = MessageSenderApiClient(
+            settings.MESSAGE_SENDER_TOKEN,
+            settings.MESSAGE_SENDER_URL,
+        )
+
+        data = self.retrieve_identity_info(is_client, msisdns)
+
+        data = self.retrieve_registration_info(is_client, data)
+
+        (data, list_length) = self.retrieve_messages(ms_client, data,
+                                                     start_date, end_date)
+
+        spreadsheet = self.populate_spreadsheet(msisdns, data, list_length)
+
         output_file = generate_random_filename()
         spreadsheet.save(output_file)
 
         task_status.status = ReportTaskStatus.SENDING
         task_status.file_size = getsize(output_file)
         task_status.save()
 
-        # TODO: Use SendEmail task to send the email
-
-    def create_spreadsheet(self, msisdns):
+        if email_recipients:
+            file_name = 'msisdn-report-%s-to-%s.xlsx' % (
+                start_date.strftime('%Y-%m-%d'),
+                end_date.strftime('%Y-%m-%d'))
+            SendEmail.apply_async(kwargs={
+                'subject': email_subject,
+                'file_name': file_name,
+                'file_location': output_file,
+                'sender': email_sender,
+                'recipients': email_recipients,
+                'task_status_id': task_status_id})
```

Comment:
Do we maybe need to update the task status to completed here at the end?

Reply:
:man_facepalming:

Reply:
I also just saw that there is a test to check that the status is completed. Is that test somehow faulty? Or is there a task status change somewhere that I'm missing?

Reply:
@rudigiesler this change happens in the send_email task

Reply:
Ah okay, but if there are no email addresses specified, then we should still mark the task as complete?

Reply:
:+1: